lib/realtime/server.ex

```diff
@@ -0,0 +1,147 @@
+defmodule Realtime.Server do
+  @moduledoc """
+  Fetches live data from RTR, and forwards it to connected clients.
+  """
+
+  use GenServer
+
+  alias Gtfs.Route
+  require Logger
+
+  defmodule Opts do
+    @type t :: %__MODULE__{
+            url: String.t(),
+            poll_delay: integer
+          }
+    @enforce_keys [
+      :url,
+      :poll_delay
+    ]
+    defstruct [
+      :url,
+      :poll_delay
+    ]
+  end
+
+  @type subscriptions :: %{optional(pid()) => [Route.id()]}
+
+  @type vehicles :: %{optional(Route.id()) => [Realtime.Vehicle.t()]}
+
+  @type state :: %{
+          url: String.t(),
+          poll_delay: integer(),
+          subscriptions: subscriptions(),
+          vehicles_timestamp: integer() | nil,
+          vehicles: vehicles()
+        }
+
+  # Client functions
+
+  @spec start_link(Opts.t()) :: GenServer.on_start()
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+  end
+
+  @spec start_mocked(Opts.t()) :: pid()
+  def start_mocked(opts) do
+    {:ok, pid} = GenServer.start_link(__MODULE__, opts)
+    pid
+  end
+
+  @spec subscribe([Route.id()], GenServer.server()) :: vehicles()
+  def subscribe(route_ids, server \\ nil) do
+    server = server || __MODULE__
+    GenServer.call(server, {:subscribe, route_ids, self()})
+  end
+
+  # GenServer callbacks
+
+  @impl true
+  def init(opts) do
+    initial_state = %{
+      url: opts.url,
+      poll_delay: opts.poll_delay,
+      subscriptions: Map.new(),
+      vehicles_timestamp: nil,
+      vehicles: Map.new()
+    }
+
+    {:ok, initial_state, {:continue, :poll}}
+  end
+
+  @impl true
+  def handle_continue(:poll, state) do
+    case fetch(state.url) do
+      {:ok, vehicles} ->
+        Process.send_after(self(), :poll, state.poll_delay)
```

Comment:
They're slightly different, in that if we fail here, it probably means we have a typo in our url or something, and want to fail to start, but if we fail in the `handle_info`, it means it's a transient failure and we should keep going.

Reply:
Why couldn't it be a transient failure the first time?

Reply:
If the first call fails, suppose it could be caused my a transient problem, but it's more likely to be a typo or something we can't recover from. Continuing to start the server and pretending its healthy would make the problem harder to spot.

And I'm not worried about this small amount of repetition.